src/client_metrics.rs

```diff
@@ -178,13 +179,85 @@ pub struct MetricSample {
     pub labels: Option<HashMap<String, String>>,
 }
 
-#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Builder)]
+impl PartialEq for MetricSample {
+    fn eq(&self, other: &Self) -> bool {
+        const EPSILON: f64 = 1e-10;
+
+        let values_equal = (self.value - other.value).abs() < EPSILON;
+
+        let labels_equal = match (&self.labels, &other.labels) {
+            (Some(self_labels), Some(other_labels)) => self_labels == other_labels,
+            (None, None) => true,
+            _ => false,
+        };
+
+        values_equal && labels_equal
+    }
+}
+
+impl Eq for MetricSample {}
+
+impl MetricSample {
+    pub fn labels_to_key(&self) -> String {
+        match &self.labels {
+            Some(labels_map) => {
+                let mut sorted_entries: Vec<(&String, &String)> = labels_map.iter().collect();
+                sorted_entries.sort_by(|a, b| a.0.cmp(b.0));
+                sorted_entries.iter()
+                    .map(|(k, v)| format!("{k}:{v}"))
+                    .collect::<Vec<String>>()
+                    .join(",")
+            }
+            None => "".to_string(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
+#[serde(rename_all = "lowercase")]
+pub enum MetricType {
+    Counter,
+    Gauge,
+    #[serde(other)]
+    Unknown,
+}
+
+impl FromStr for MetricType {
+    type Err = ();
```

Comment:
It's not a big deal but this is idiomatic

```suggestion
    type Err = std::convert::Infallible;
```

Reply:
TIL. Weird that none of the materials I studied mentioned that